snuba/datasets/plans/storage_builder.py

```diff
@@ -92,112 +98,78 @@ def get_query_data_source(
-                    final=query.get_final(),
-                    sampling_rate=query.get_sample(),
-                )
-            )
-
-        cluster = self.__storage.get_cluster()
-
-        db_query_processors = [
-            *self.__storage.get_query_processors(),
-            *self.__post_processors,
-            MandatoryConditionApplier(),
-            MandatoryConditionEnforcer(
-                self.__storage.get_mandatory_condition_checkers()
-            ),
-        ]
-
-        return [
-            ClickhouseQueryPlan(
-                query=clickhouse_query,
-                plan_query_processors=[],
-                db_query_processors=db_query_processors,
-                storage_set_key=self.__storage.get_storage_set_key(),
-                execution_strategy=SimpleQueryPlanExecutionStrategy(
-                    cluster=cluster,
-                    db_query_processors=db_query_processors,
-                    splitters=self.__storage.get_query_splitters(),
-                ),
-            )
-        ]
-
-
-class SelectedStorageQueryPlanBuilder(ClickhouseQueryPlanBuilder):
+class StorageQueryPlanBuilder(ClickhouseQueryPlanBuilder):
     """
-    A query plan builder that selects one of multiple storages in the dataset.
+    Builds the Clickhouse Query Execution Plan for a dataset which supports single storages,
+    multiple storages, unsliced storages, and sliced storages.
     """
 
     def __init__(
         self,
-        selector: QueryStorageSelector,
+        storages: List[StorageAndMappers],
+        selector: Optional[QueryStorageSelector] = None,
         post_processors: Optional[Sequence[ClickhouseQueryProcessor]] = None,
+        partition_key_column_name: Optional[str] = None,
     ) -> None:
+        self.__storages = storages
         self.__selector = selector
         self.__post_processors = post_processors or []
+        self.__partition_key_column_name = partition_key_column_name
 
     @with_span()
     def build_and_rank_plans(
         self, query: LogicalQuery, settings: QuerySettings
     ) -> Sequence[ClickhouseQueryPlan]:
-        with sentry_sdk.start_span(
-            op="build_plan.selected_storage", description="select_storage"
-        ):
-            storage, mappers = self.__selector.select_storage(query, settings)
+        if len(self.__storages) < 1:
+            raise QueryStorageSelectorError("No storages specified to select from.")
+
+        if not self.__selector:
+            # Default to the first and only storage and mapper
+            if len(self.__storages) == 1:
+                storage, mappers, _ = self.__storages[0]
+            else:
+                # If there are both readable and writable storages, select the readable one.
+                # Multiple writable storages are not supported.
+                readable_storages = [
+                    storage for storage in self.__storages if not storage.is_writable
```

Comment:
Related to https://github.com/getsentry/snuba/pull/3509#discussion_r1047529358

I think the problem is that this function is trying to be too generic and handle too many cases. There are currently three use cases for storage selection:
1) Single storage: The entity provides a single storage and mappers, and those are always used.
2) Selected storage: The entity provides a class which selects a storage/mappers. <- the problem case
3) Sliced storage: The entity provides a single storage and mappers, and some extra fields that allow choosing a cluster. Note that this is the same as the Single Storage, but changes how the cluster is fetched.

This logic only needs to allow and handle those three cases explicitly. So `StorageQueryPlanBuilder` would be initialized the same way you have now.

Pick a storage:
- Does the builder specify a selector? Pass the list of storages to the selector and get back a single `StorageAndMappers`. <- the major change.
- Otherwise, use the provided storage/mapper.

Fetch the cluster:
- Does the builder have sliced storage fields? Use those to fetch the cluster.
- Otherwise, get the cluster from the storage selected above.

The major change here is to have the storage selectors themselves deal with choosing the storage.

Reply:
The main function can definitely be separated to draw a clearer distinction as Volo mentioned. This PR attempts to cover those three main cases explicitly, but also adds some functionality in preparation for defining `StorageAndMappers` in the entity https://github.com/getsentry/snuba/pull/3509#discussion_r1047804306.

For that functionality to work, we wouldn't be able to "Otherwise, use the provided storage/mapper"  which is why the logic for defaulting to a readable storage is there.